src/vm-memory/src/mmap.rs

```diff
@@ -460,6 +460,56 @@ impl GuestMemoryMmap {
     pub fn is_dirty_tracking_enabled(&self) -> bool {
         self.regions.iter().all(|r| r.dirty_bitmap().is_some())
     }
+
+    pub fn read_from<F>(
+        &self,
+        addr: GuestAddress,
+        src: &mut F,
+        count: usize,
+    ) -> result::Result<usize, vm_memory_upstream::guest_memory::Error>
+    where
+        F: Read,
+    {
+        self.try_access(
+            count,
+            addr,
+            |offset,
+             len,
+             caddr,
+             region|
+             -> result::Result<usize, vm_memory_upstream::guest_memory::Error> {
+                // Check if something bad happened before doing unsafe things.
+                assert!(offset <= count);
+                let len = std::cmp::min(len, 4096);
```

Comment:
Why do we do this ?

Reply:
We don't. I will remove this.

Reply:
More precisely, I meant why do we need this custom logic ? I see that the buffer len is equal to the page len . But why ? Can't we use the default GuestMemory implementation of this function ? I think the dirty tracking should work correctly anyway.

Reply:
This custom logic is a part of the upstream implementation. The `read_from` function in upstream has two workflow branches: one for the regions which have `as_mut_slice` implemented and another one otherwise. 
For the second case, the content of the source (the object we want to read from) is read into a buffer and then the buffer is written in the region (which is what we want, because the region `write` function owns logic to mark the pages dirty).  
For the first case, the upstream uses a slice of the region and writes there directly from source by calling `read` on the object. 

Because of vsock, our local region has an implementation of `as_mut_slice`, so it matches this first case. Therefore, by using the upstream implementation, we would not be able to mark pages as dirty. To solve this issue, I brought locally the part we need from upstream.

Reply:
Ok, I understand the issue now. Thanks for the explanation.